### DISCUSSION TITLE
Undo/Redo not jumping to what was undid or redone

### DISCUSSION THREAD

**Discussion by zambetti** [zed-industries/zed#6626] [Feature Requests]
When I undo/redo, sometimes (I don't know why) the buffers don't jump to the line to show me what was undid or redone.

Weirdly, if I have the same file open in different buffers, it seems like they each will jump differently when I undo/redo. Perhaps they will jump if that is the buffer that that particular change was performed in?

Undo/redo should always be shown somehow; to perform changes without displaying them, is unnerving.

Likely related to this, the menubar options Edit > Undo, and Edit > Redo are sometimes grayed-out (indicating that there is nothing to undo/redo), but in fact, when you use the key commands, undos/redos are performed.

Has anyone else encountered these oddities with undo/redo?

**Comment by maxbrunsfeld** [comment]
Thanks for bringing this up. I wrote a bug report this with reproduction steps. https://github.com/zed-industries/community/issues/2061, for the scrolling issue that you described.

**Reply by jpike88** [reply]
Hi Max, this is rapidly becoming a major point of frustration for me, I'm losing count of when I want to undo something and get confused where the undo actually took place. Can this be escalated, considering how fundamental a UI bug this is?

**Comment by zambetti** [comment]
Thank you Max. From a UI perspective, two solutions come to mind to solve this.

1) Show the undo/redo in whichever currently-visible pane is nearest to the relevant line. This would best preserve the scroll-positions in split-editing scenarios (where the same file is open in two panes simultaneously at different scroll-positions). However, it could be disorienting to see multiple buffers jumping scroll-position on undo/redo.

2) Only jump in the pane that is currently focused. This would likely be easiest to implement, and should feel familiar since this is how most editors work. However, in the split-editing scenario, the scroll-position is lost due to the undo-redo jump.

Both of these solutions would handle the case where changes were made in a pane that was later closed. They would also both avoid non-visible pane tabs popping into focus unexpectedly just to show undos/redos.

Considering that with all the navigation features it's pretty easy to return to where you were; #2 might be best since it's most familiar.

**Comment by viniciusnevescosta** [comment]
I wanted to share a very stressful situation. I just lost an entire file of code because I right clicked on another file name. For some reason the editor selects 2 files (the one I was editing and the one I selected), and this behavior doesn't make sense, if I wanted to select 2 files I would have simply pressed shift or cmd and then clicked on the files. I ended up losing an empty file that I wanted and what I was editing. I'm sorry, but this should be a basic function of an IDE.

**Reply by jpike88** [reply]
@notpeter I think the way everything's just left to discussions and not handled like a proper bug report is detrimental to Zed's quality. What do you do when something like this is reported in a discussion? Does anyone in the team even know?

**Comment by KennyLindahl** [comment]
Bump, as this is one of the most basic and essential features that is needed.

**Reply by KennyLindahl** [reply]
Mostly seems to work now, but still happens sometimes.
It would also be good if it first jumps to the line that is out of screen (no revert unless you actually see the change happening)

**Comment by 5eh** [comment]
Any fixes here? Still getting this annoying UI issue.

**Comment by waleed1892** [comment]
Any update on fixing this?

**Reply by jpike88** [reply]
Can you describe exactly how the problem occurs? What specific action you did? Was it just typing something, or hitting tab from dit prediction etc?

**Reply by waleed1892** [reply]
I'm on MacOS and doing undo using cmd +z.
The code is reverted while staying on same line. It should jump cursor to the line where undo is happening.

**Reply by jpike88** [reply]
Ok, but what did you do before hitting undo? Was it tab from an edit prediction, just typing a letter, hitting save and auto-format happening too...

**Reply by jpike88** [reply]
@waleed1892 it's important information, just saying undo doesn't work isn't going to help. There are some edge cases, and some fixes already applied.

https://github.com/zed-industries/zed/issues/29652

See this one I opened today

**Comment by JosephTLyons** [comment]
Thanks for reaching out. We want to better support our community, and the best way for us to do that is through our official Discord. The team is active there and happy to help.

We’re closing the “Help and General Q&A” category on GitHub Discussions and directing all support questions to Discord going forward.

https://zed.dev/community-links
